mat64/io.go

```diff
@@ -5,14 +5,152 @@
 package mat64
 
 import (
+	"bytes"
 	"encoding/binary"
+
+	"github.com/gonum/matrix"
 )
 
 var (
-	littleEndian  = binary.LittleEndian
-	bigEndian     = binary.BigEndian
-	defaultEndian = littleEndian
-
+	binLE       = binary.LittleEndian
 	sizeInt64   = binary.Size(int64(0))
 	sizeFloat64 = binary.Size(float64(0))
 )
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Dense is little-endian encoded as follows:
+//   0 -  8  number of rows    (int64)
+//   8 - 16  number of columns (int64)
+//  16 - ..  matrix data elements (float64)
+//           [0,0] [0,1] ... [0,ncols-1]
+//           [1,0] [1,1] ... [1,ncols-1]
+//           ...
+//           [nrows-1,0] ... [nrows-1,ncols-1]
+func (m Dense) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(make([]byte, 0, m.mat.Rows*m.mat.Cols*sizeFloat64+2*sizeInt64))
+	err := binary.Write(buf, binLE, int64(m.mat.Rows))
+	if err != nil {
+		return nil, err
+	}
+	err = binary.Write(buf, binLE, int64(m.mat.Cols))
+	if err != nil {
+		return nil, err
+	}
+
+	nrows, ncols := m.Dims()
+	for i := 0; i < nrows; i++ {
+		for j := 0; j < ncols; j++ {
+			err = binary.Write(buf, binLE, m.at(i, j))
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes the binary form into the receiver.
+// It panics if the receiver is a non-zero Dense matrix.
+//
+// See MarshalBinary for the on-disk layout.
+func (m *Dense) UnmarshalBinary(data []byte) error {
+	if !m.isZero() {
+		panic("mat64: unmarshal into non-zero matrix")
+	}
+
+	buf := bytes.NewReader(data)
+	var rows int64
+	err := binary.Read(buf, binLE, &rows)
+	if err != nil {
+		return err
+	}
+	var cols int64
+	err = binary.Read(buf, binLE, &cols)
+	if err != nil {
+		return err
+	}
+
+	m.mat.Rows = int(rows)
+	m.mat.Cols = int(cols)
+	m.mat.Stride = int(cols)
+	m.capRows = int(rows)
+	m.capCols = int(cols)
+
+	// maxLen is the biggest slice/array len one can create on a 32/64b platform.
+	const maxLen = int64(int(^uint(0) >> 1))
+
+	size := rows * cols
+	if rows < 0 || cols < 0 || int(size) < 0 || size > maxLen {
+		return matrix.ErrShape
```

Comment:
I don't think this is ErrShape, you can define a private error along the lines of "mat64: too big".